src/actor/typst.rs

```diff
@@ -98,44 +140,98 @@ impl TypstActor {
                 panic!();
             }
         };
+        let (pre_doc_sender, pre_doc_receiver) = mpsc::unbounded_channel();
 
+        let doc_receiver = doc_sender.subscribe();
         Self {
             compiler_driver,
             fs_watcher,
             mailbox,
-            doc_sender,
-            renderer_sender,
+            pre_doc_sender,
+            pre_doc_receiver: Some(pre_doc_receiver),
+            doc_sender: Some(doc_sender),
+            doc_receiver,
+            renderer_sender: Some(renderer_sender),
             doc_to_src_jump_sender,
             src_to_doc_jump_sender,
         }
     }
     pub fn run(mut self) {
-        self.compile();
-        let Ok(_) = self
-            .fs_watcher
-            .watch(&self.compiler_driver.world.root, RecursiveMode::Recursive)
-        else {
-            error!("TypstActor: failed to watch filesystem events");
+        self.compile(None);
+        let root = self.compiler_driver.world.root.clone();
+        if let Err(e) = self.fs_watcher.watch(&root, RecursiveMode::Recursive) {
+            error!("TypstActor: failed to watch filesystem events: {e}");
             panic!();
         };
+
+        let mut pre_doc_receiver = self.pre_doc_receiver.take().unwrap();
+        let doc_sender = self.doc_sender.take().unwrap();
+        let renderer_sender = self.renderer_sender.take().unwrap();
+        std::thread::Builder::new()
+            .name("pre_receiver".to_owned())
+            .spawn(|| {
+                tokio::runtime::Builder::new_current_thread()
```

Comment:
~~pls don't build a new runtime or you have very decent reason to do this~~ This seems to be OK